stix_shifter_modules/msatp/stix_transmission/connector.py

```diff
@@ -1,29 +1,198 @@
+            device_registry_events_data.pop(k)
+        else:
+            registry_value_dict[k] = ''
+    device_registry_events_data["RegistryValues"] = [registry_value_dict]
+
+
+def organize_ips(data):
+    ips_comp_lst = data.pop('IPAddressesSet')
+    flat_lst = [ip_obj['IPAddress'] for ips in ips_comp_lst for ip_obj in json.loads(ips) if 'IPAddress' in ip_obj]
+    data['IPAddresses'] = flat_lst
+
+
+def create_event_link(data, timestamp):
+    try:
+        if 'DeviceId' in data:
+            # parse timestamp to date object striping nanoseconds
+            timestamp_dt = datetime.strptime(timestamp[:-9], "%Y-%m-%dT%H:%M:%S")
+            timeline_start = (timestamp_dt - timedelta(seconds=1)).strftime("%Y-%m-%dT%H:%M:%S") + ".000Z"
+            timeline_end = (timestamp_dt + timedelta(seconds=1)).strftime("%Y-%m-%dT%H:%M:%S") + ".000Z"
+            event_link = 'https://%s/machines/%s/timeline?from=%s&to=%s' % (
+                'security.microsoft.com', data.get('DeviceId'), timeline_start, timeline_end)
+            data['event_link'] = event_link
+    except Exception as ex:
+        data['event_link'] = ''
+
+
+def remove_duplicate_ips(event_data):
+    ## remove duplicate ips between LocalIP, PublicIP and IPAddresses:
+    if 'PublicIP' in event_data and 'LocalIP' in event_data and \
+            event_data['PublicIP'] == event_data['LocalIP']:
+        event_data.pop('PublicIP')
+    remove_duplicate_ips_from(event_data, 'LocalIP')
+    remove_duplicate_ips_from(event_data, 'PublicIP')
+
+
+def remove_duplicate_ips_from(event_data, prop_name):
+    if 'IPAddresses' in event_data and prop_name in event_data:
+        filtered = [x for x in event_data['IPAddresses'] if
+                    not x == event_data[prop_name]]
+        if len(filtered) > 0:
+            event_data['IPAddresses'] = filtered
+        else:
+            event_data.pop('IPAddresses')
+
+
+def unify_alert_fields(event_data):
+    techniques_lists = []
+    # attack techinques is a string due to the make_list function in the kql
+    # need to convert it back to dict
+    if 'AttackTechniques' in event_data:
+        for techniques_lst in event_data['AttackTechniques']:
+            try:
+                attack_techniques = json.loads(techniques_lst)
+            except json.decoder.JSONDecodeError:
+                attack_techniques = ''
+            finally:
+                techniques_lists.append(attack_techniques)
+        event_data['AttackTechniques'] = techniques_lists
+
+    alerts = []
+    alerts_count = len(event_data['AlertId'])
+    for i in range(alerts_count):
+        alert_dct = {k: (event_data[k][i] if len(event_data[k]) > i else '')
+                     for k in Connector.ALERT_FIELDS if k in event_data}
+        if alert_dct['AlertId'] not in [alert['AlertId'] for alert in alerts]:
+            alerts.append(alert_dct)
+    event_data['Alerts'] = json.dumps(alerts)
+    for f in Connector.ALERT_FIELDS:
+        event_data.pop(f, None)
 
```

Comment:
Is it possible to put all these functions in a utility class without crowding the Connector() class? Its for the better maintainability. OR `create_results_connection` can be separated into a `ResultsConnector` class (/stix_transmission/results_connector.py)

Reply:
ok, done. all logic was moved to connector_post_processing.py and out of connector.py